### PR TITLE
Implement logic for storing fields that are neither dimensions nor metrics (aka tags)

### PR DESCRIPTION
**Pull request by salvatore-campagna**
We want labels to be collected storing just the latest value
when aggregating documents in a rollup index. This means
labels will behave similarly to counter metrics.

### CODE REVIEW

**Comment by salvatore-campagna** [comment]
elasticsearch-ci/part-2 is failing at the moment for a mapping issue involving IPv4 addresses which are handled as keywords. The following line ```return format.format(values.nextValue());``` is failing because the byte array includes non-UTF8 characters like '0xFF' which are used to encode an IPv4 address.

**Comment by salvatore-campagna** [comment]
I will see if I can completely remove all the `FieldCapabilities` stuff. Now that we use the original index mapping it should be possible.

**Comment by salvatore-campagna** [comment]
I decided to keep the logic using field capabilities to separate different fields (dimensions, metrics and labels) because after writing the code not using field capabilities I didn't like it. When identifying labels we need to take into account different things like if the field is "aggregatable" or not or if the field is a text field or a metadata field. Tese checks are easier using field capabilities rather than using mappings. That logic is used, anyway, just at the very beginning of the rollup process to separate fields. Other parts of the code work using mappings.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java
Comment (salvatore-campagna) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java:302]: This is incorrect. For multi value fields it collects `docValueCount` times the same (first) value. Call to next values needs to be in the loop.

**Comment by salvatore-campagna** [comment]
I will get rid of all that mapping "custom" logic and replace it with whatever service/api we have.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
I finally managed to get rid of `FieldCapabilities`. I also had to fix an issue caused by the rounding applied to the @timestamp field while creating the rollup document. It could happen, indeed, that after the @timestamp is calculated, due to rounding it down to the boundary of the bucket interval, the resulting value is smaller than the `index.time_series.start_time. That caused some documents not to be indexed.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by salvatore-campagna** [comment]
I think I will add a test using an unmapped field as a label and check if propagating the latest value works correctly in that case too.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilities.java
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilities.java:387]: I don't think you use this any more, right?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/action/RollupIndexerAction.java
Comment (nik9000) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/action/RollupIndexerAction.java:69]: :+1:


I think we don't need bwc here because this is behind a feature flag.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/AbstractFieldProducer.java
Comment (nik9000) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/AbstractFieldProducer.java:10]: Something like this deserves javadoc I think. Also, can it be package private? It looks like something specific to rollup v2 and package private just gives the extra context that "this is contained".

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java
Comment (salvatore-campagna) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:216]: I am looking at whether I can extract some of this functionality into an helper class or something so that we can write
```
if (helper.isDimension(field)) {
...
} else if (helper.isMetric(field) {
...
} else if (helper.isLabel(field) {
...
}
```

This way we can reuse (and unit test) it.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/LabelFieldProducer.java
Comment (salvatore-campagna) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/LabelFieldProducer.java:79]: At the moment this is the only strategy used for labels. It was decided to just use the latest value to make things easier to start with. It is not difficult to imagine that other strategies might be ' FirstValueLabel' or `LatestNValues` or whatever...even though, according to my understanding this has never been discussed.
The idea is that we have a bunch of values but we need to pick one of them to write in the rollup index.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/Collectable.java
Comment (romseygeek) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/Collectable.java:15]: Do we actually use this interface anywhere? It's implemented by the FieldProducers but we never refer to anything as a Collectable so I think we can probably just fold it into AbstractFieldProducer

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/part-1

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java
Comment (csoulios) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java:250]: I see that you pass the failure handler `listener:onFailure` all the way down to the `addMetricFields()` method.
Also, you catch the `IOException` and call the failure handler here. 

I think you should choose one of the two approaches here. Either throw exception in the calling methods and wrap it in the failure handler at the top level or use the failure handler wherever an exception is thrown. 

Since methods `createRollupIndexMapping()` and `addMetricFields()` are called synchronously, you can signal errors by throwing exceptions. The exceptions will be called and delegated to the failure handler at the top level. I think this will simplify testing the methods signatures a bit and make testing easier.

**Review by csoulios — COMMENTED** [review]
I have gone through the code. I am very happy with how this forms. 

I have added some comments

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/LabelFieldProducer.java
Comment (salvatore-campagna) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/LabelFieldProducer.java:79]: Well there is a difference...the metric one has T = Number while for a label T = Object.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java
Comment (salvatore-campagna) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java:314]: This is done to handle multi-vaule fields...if I don't do this I need to pass the `docValueCount` and have to change the `collect` method signature.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/FieldTypeHelper.java
Comment (salvatore-campagna) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/FieldTypeHelper.java:12]: This interface exists because I dont want to leak the actual class implementation when I do this:

```
public FieldTypeHelper timeseriesHelper(final String timestampField) throws IOException {
            final MapperService mapperService = indicesService.createIndexMapperServiceForValidation(indexMetadata);
            final CompressedXContent sourceIndexCompressedXContent = new CompressedXContent(indexMapping);
            mapperService.merge(MapperService.SINGLE_MAPPING_NAME, sourceIndexCompressedXContent, MapperService.MergeReason.INDEX_TEMPLATE);
            return new TimeseriesFieldTypeHelper(mapperService, timestampField);
        }
```

So it is there to avoid the client knowing which actual class implementation we are using.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/FieldTypeHelpers.java
Comment (salvatore-campagna) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/FieldTypeHelpers.java:18]: Here I am using (an old) convention that I have seen used multiple times in java code. Using a class that is called <something>s as a collector os static resources (classes, methods) for a thing called <something>. I can of course use `TimeSeriesFieldTypeHelper` but then I need to make that class public.

File: x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/rollup/v2/RollupActionSingleNodeTests.java
Comment (csoulios) [inline x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/rollup/v2/RollupActionSingleNodeTests.java:510]: This method is becoming too long. Is it worth breaking into separate steps such as `assertRollupSettings`, `assertRollupAggs`, `assertRollupMapping`. And then we can have `assertRollupIndex` call those methods

File: x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/rollup/v2/RollupActionSingleNodeTests.java
Comment (csoulios) [inline x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/rollup/v2/RollupActionSingleNodeTests.java:199]: Thanks for fixing this!

**Comment by salvatore-campagna** [comment]
@elasticmachine test this please

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Review by csoulios — COMMENTED** [review]
Thanks for iterating on this. 
I left some more comments.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TimeseriesFieldTypeHelper.java
Comment (salvatore-campagna) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TimeseriesFieldTypeHelper.java:38]: This is basically to avoid fields of type `text` which is not aggregatable...while it would work with fields of type `keyword` which are aggregatable.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java
Comment (salvatore-campagna) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java:386]: Not really, because I think that needs to decide first how do we handle that which I don't have any idea at the moment.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TimeseriesFieldTypeHelper.java
Comment (salvatore-campagna) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TimeseriesFieldTypeHelper.java:38]: If I remove that condition the tests results in an errors saying something like "you need to use aggregatable fields or use <some API I don't remember> to support non-aggregatable fields

Comment (csoulios) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TimeseriesFieldTypeHelper.java:38]: cool thanks for the explanation. I had missed that

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TimeseriesFieldTypeHelper.java
Comment (csoulios) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TimeseriesFieldTypeHelper.java:38]: This actually means that we cannot have a `text` field as label?

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java
Comment (csoulios) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java:402]: I think it should work in the same way as aggregations work on multi-value fields.

**Review by csoulios — APPROVED** [review]
LGTM! 

Thank you for iterating on this

**Comment by salvatore-campagna** [comment]
> LGTM!
> 
> Thank you for iterating on this

Are we still waiting for something before being able to merge or can I proceed?

**Comment by csoulios** [comment]
> Are we still waiting for something before being able to merge or can I proceed?

Please go ahead and merge it. 

We have identified 2 TODO tasks that can be implemented in follow-up PRs:
1. `text` fields as labels
2. Computing metric aggs on array fields

**Comment by salvatore-campagna** [comment]
> > Are we still waiting for something before being able to merge or can I proceed?
> 
> Please go ahead and merge it.
> 
> We have identified 2 TODO tasks that can be implemented in follow-up PRs:
> 
> 1. `text` fields as labels
> 2. Computing metric aggs on array fields

I will create two tasks in https://github.com/elastic/elasticsearch/issues/74660 to track them.

**Pull request merged**
